src/remark/code-block-snippets.js

```diff
@@ -0,0 +1,125 @@
+                // From-to range of line numbers.
+                let range = [];
+                // XXX: Online reading line-by-line from buffer causes some async JS abnormalities.
+                const codeSplit = code.split("\n");
+                code = "";
+
+                const linesRegex = /L(\d+)(-L(\d+))?/;
+                if (linesRegex.test(anchor)) {
+                    // Anchor contains a single line number (e.g. L14) or range (e.g. L10-L23).
+                    linesParts = linesRegex.exec(anchor);
+                    range.push(linesParts[1]);
+                    range.push(linesParts[3] ?? linesParts[1]);
+
+                    let lineN = 1
+                    for (const line of codeSplit) {
+                        if (lineN >= range[0]) {
+                            if (lineN <= range[1]) {
+                                code += (line + "\n");
+                            } else {
+                                break;
+                            }
+                        }
+                        lineN++;
+                    }
+
+                    // Sanity check.
+                    if ((code.match(/\n/g) || []).length != range[1]-range[0]+1) {
+                        throw new Error(`line range ${range[0]}:${range[1]} not valid in file ${sourceFilePath}`);
+                    }
+                } else {
+                    // Anchor references region with #region and #endregion literals.
+                    const regionRegex = /.*#region (.*)/;
+                    const endRegionRegex = /.*#endregion (.*)/;
+
+                    let write = false;
+                    let lineN = 1;
+                    for (const line of code.split("\n")) {
+                        const region = regionRegex.exec(line);
+                        const endRegion = endRegionRegex.exec(line);
+
+                        if (region && region[1] == anchor) {
+                            range.push(lineN+1);
+                            write = true;
+                        } else if (endRegion && endRegion[1] == anchor) {
+                            range.push(lineN-1);
+                            break;
+                        } else if (write) {
+                            code += (line + "\n");
+                        }
+
+                        lineN++;
+                    }
+
+                    // Sanity check.
+                    if (range.length != 2) {
+                        throw new Error(`region '${urlAnchor[1]}' not found in ${sourceFilePath}`);
+                    }
+                }
+
+                // Append line numbers to source URL even if region name was initially provided.
+                node.url += `#L${range[0]}`
+                if (range[1]) {
+                    node.url += `-L${range[1]}`;
+                }
+            }
+
+            // Remove trailing newline, if any.
+            code = code.replace(/\n$/, "")
+            node.value = code;
+        }
```

Comment:
I'd refactor to:
```js
        const code = fs.readFileSync(sourceFilePath, 'utf-8');
        node.url = viewLinkUrl(sourceFilePath);

        // Remove trailing newline, if any.
        let snippet = code.replace(/\n$/, "")

        // Check, if anchor matches line number or region name inside the file.
        if (anchor) {
            // From-to range of line numbers.
            let range;
            const linesParts = /L(\d+)(-L(\d+))?/.exec(anchor)
            if (linesParts) {
                // Anchor contains a single line number (e.g. L14) or range (e.g. L10-L23).
                range = [
                    parseInt(linesParts[1]),
                    parseInt(linesParts[3] ?? linesParts[1]),
                ]
            } else {
                // Anchor references region with #region and #endregion literals.

                // +1/-1 to exclude lines with #region, and +1 for 1-based numbering
                range = [
                    code.split('\n').findIndex((line) => line.includes(`#region ${anchor}`)) + 1 + 1,
                    code.split('\n').findIndex((line) => line.includes(`#endregion ${anchor}`)) - 1 + 1,
                ]
                if (range[0] < 2) throw new Error(`"#region ${anchor}" not found`)
                if (range[1] < 0) throw new Error(`"#endregion ${anchor}" not found`)
                if (range[0] > range[1]) throw new Error(`"#region ${anchor}" appears after "#endregion ${anchor}"`)
            }

            // Append line numbers to source URL even if region name was initially provided.
            node.url += `#L${range[0]}-L${range[1]}`;

            // -1 for 1-based numbering
            snippet = code.split('\n').slice(range[0] - 1, range[1] - 1 + 1).join('\n')
        }

        node.value = snippet;
```

Reply:
Thanks! I'm not fond of re-reading the file twice, when region names provided, but it won't give big of an impact for real world examples since they are small.